lib/client/api.go

```diff
@@ -2724,13 +2766,75 @@ func (tc *TeleportClient) runCommandOnNodes(ctx context.Context, clt *ClusterCli
 			}
 			defer nodeClient.Close()
 
-			fmt.Printf("Running command on %v:\n", nodeName(address))
+			displayName := nodeName(node.addr)
+			if node.hostname != "" {
+				displayName = node.hostname
+			}
+			fmt.Printf("Running command on %v:\n", displayName)
 
-			return trace.Wrap(nodeClient.RunCommand(ctx, command))
+			if err := nodeClient.RunCommand(ctx, command); err != nil && tc.ExitStatus == 0 {
+				return trace.Wrap(err)
+			}
+			resultsCh <- execResult{
+				hostname:   displayName,
+				exitStatus: tc.ExitStatus,
+			}
+			return nil
 		})
 	}
 
-	return trace.Wrap(g.Wait())
+	if err := g.Wait(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	close(resultsCh)
+	results := make([]execResult, 0, len(resultsCh))
+	for result := range resultsCh {
+		results = append(results, result)
+	}
+
+	return trace.Wrap(tc.writeCommandResults(results))
+}
+
+func (tc *TeleportClient) writeCommandResults(nodes []execResult) error {
+	var successFile *os.File
+	var failFile *os.File
+	var err error
+	if tc.SSHLogDir != "" {
+		successFile, err = os.Create(filepath.Join(tc.SSHLogDir, "hosts.succeeded"))
```

Comment:
Do we always need to have both files? In most cases I would expect empty hosts.failed. Maybe delete a file if we know there's nothing in there (depending on ratio of failed nodes to total nodes)?